### PR TITLE
[WebGPU] WGSL compiler should support unicode characters

### PR DESCRIPTION
**Pull request by tadeuzagallo**
#### 087346e089cf2fe3c75037da7eb17553534a6fdb
<pre>
[WebGPU] WGSL compiler should support unicode characters
<a href="https://bugs.webkit.org/show_bug.cgi?id=266774">https://bugs.webkit.org/show_bug.cgi?id=266774</a>
<a href="https://rdar.apple.com/120004379">rdar://120004379</a>

Reviewed by Mike Wyrzykowski and Yusuke Suzuki.

For the most part, we already supported unicode, but the implementation for
isIdentifierStart and isIdentifierContinue needed to be updated. After this patch
all the identifier tests in the CTS (under shader,validation,parse,identifiers:*)
are passing.

* Source/WebGPU/WGSL/Lexer.cpp:
(WGSL::isIdentifierStart):
(WGSL::isIdentifierContinue):
(WGSL::Lexer&lt;T&gt;::nextToken):
* Source/WebGPU/WGSL/Lexer.h:
(WGSL::Lexer::isIdentifierStart): Deleted.
(WGSL::Lexer::isIdentifierContinue): Deleted.
* Source/WebGPU/WGSL/TypeCheck.cpp:
(WGSL::TypeChecker::check):
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/ShaderModule.mm:
(WebGPU::Device::createShaderModule):

Canonical link: <a href="https://commits.webkit.org/277108@main">https://commits.webkit.org/277108@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/ba5783891385f85fa23c829fb3f6035757e42459

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46579 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25736 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49186 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49256 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42624 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/30096 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23198 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/37982 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47158 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22724 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40155 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19247 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/20097 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/41271 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4624 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/42893 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41650 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51138 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/21586 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/17992 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45272 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/22876 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44224 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10321 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/23313 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22579 "Built successfully") | | |

### CODE REVIEW

**Review by mwyrzykowski — APPROVED** [review]

File: Source/WebGPU/WGSL/Lexer.cpp [hunk 1]
```diff
@@ -424,6 +463,11 @@ FOREACH_KEYWORD(MAPPING_ENTRY)
             if (UNLIKELY(reservedWordSet.contains(view)))
                 return makeToken(TokenType::ReservedWord);
 
+
+            if (UNLIKELY(length >= 2 && *startOfToken == '_' && *(startOfToken + 1) == '_'))
```
Comment (Constellation) [inline Source/WebGPU/WGSL/Lexer.cpp:467]: In JavaScript, we also allow `$`. Does WGSL allow it?

File: Source/WebGPU/WGSL/Lexer.cpp [hunk 2]
```diff
@@ -35,6 +35,41 @@
...
+
+static unsigned isIdentifierStart(LChar character, const LChar*, const LChar*)
+{
+    return isASCIIAlpha(character) || character == '_';
+}
+
+static unsigned isIdentifierContinue(LChar character, const LChar*, const LChar*)
+{
+    return isASCIIAlphanumeric(character) || character == '_';
+}
```
Comment (Constellation) [inline Source/WebGPU/WGSL/Lexer.cpp:71]: Ditto.

**Review by Constellation — APPROVED** [review]
r=me

File: Source/WebGPU/WGSL/Lexer.cpp (hunk 1 above)
Comment (tadeuzagallo) [reply Source/WebGPU/WGSL/Lexer.cpp:467]: I don't know if `XID_Start` includes it, but the spec only specifies `XID_Start` + `_`: https://www.w3.org/TR/WGSL/#identifiers

**Pull request merged**